versioning/versioning.rb

```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env ruby
+
+GIT_VERSION_COMMAND='git describe --tags --abbrev=8 --dirty 2> /dev/null'
+# Based on https://semver.org/#semantic-versioning-200 but allows `v` in front to match git tags
+GIT_SEMVER_REGEX=/^v?(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/
+
+class Versioning
+  class << self
+    def verify_git!
+      unless git_exists?
+        raise('The command `git` does not exist!')
+      end
+
+      unless git_dir?
+        raise("The current directory `#{Dir.pwd}` is not a git tree!")
+      end
+    end
+
+    def verify_semver_tags!
+      unless `#{GIT_VERSION_COMMAND}`.strip =~ GIT_SEMVER_REGEX
```

Comment:
We don't want to accepts tags that include `-<pre-release>` or `+<build>` sequences, so we really should check against `v<version core>` only.

Reply:
I think the version script should also work if a project has alpha or beta tags. @viovanov what would you prefer?

Reply:
I think we should include the pre release or meta pieces. @jandubois what's your thinking for disallowing?

Reply:
> what's your thinking for disallowing?

Mostly that it makes it harder to create the assembled semver. We can no longer rely on `git describe` alone for this:

```
$ g tag v2.4.0+gold HEAD^
$ g describe --tags --dirty
v2.4.0+gold-5-gcbc89373
```

If you have a tag with a `+<build>` element, the number of commits, and the commit id become part of that `<build identifier>`, in the example above `gold-5-gcbc89373`. Instead we would need to inject our pre-release id in the middle, like `v2.4.0-5-gcbc89373+gold`.

Even when the tag has just a `-<pre-release>` element, we have to decide if we want to append or prepend, and we probably should use a `.`, as in `v2.4.0-alpha.5-gcbc89373`, so we end up with 2 `<pre-release identifier` elements: `alpha`, and `5-gcbc89373`.

I wanted to avoid thinking about all this, and having to make decisions. Making it throw an error initially means we can later implement it however we want without breaking compatibility.

After thinking about the release process some more, I do see the utility of including pre-release identifiers in the tags. So my suggestion would be:

* Allow the tag to have an optional `-<pre-release>` element
* Disallow the tag from having a `+<build>` element.
* We attach the `<number of commits>-g<commit id>` as a separate `<pre-release identifier>`, meaning we join it with a `.` instead of a `-` if the tag already has a `<pre-release` part.

This still allows us to use the `git describe` output as-is. If it includes a `+` character, throw an error. If there is a dash **and** a commit pre-release id, change the penultimate dash into a dot:

```
version.gsub!(/(.*-.*)-(\d+-g[0-9a-zA-Z]+)$/, '\1.\2')
```

I'm willing to change my mind on forbidding the `+<build>` part after I learn why this would be needed/useful. I would think the build metadata would be added by the build system and not be derived from a git tag.

Reply:
@jandubois I'm ok with that - I don't really like meta information anyway. The `+` sign has always been a source of problems.

Reply:
@jandubois +1 to all your suggestions, except the `.`. I think we should just leave the `-` in, since we don't know how the existing element looks like.

Reply:
> I think we should just leave the `-` in, since we don't know how the existing element looks like.

It is not about how it looks; it is about the semantics. The `-<pre-release>` element is a list of identifiers, separated by dots. Dashes are part of identifiers (except for the initial one that is just the marker that the following characters are the `<pre-release>` list).

The `+<build>` element works the same way: it is a list of identifiers separated by dots, and dashes are just part of the identifiers. That's also why the `-<pre-release>` element has to come first; once you are inside `<build>`, the dash is just an `<identifier character>` and cannot signal the beginning of a different element.

Maybe it becomes clearer when you look at the parsed versions.

`v2.4.0-alpha-5-gcbc89373` means:

```json
{
  "version_core": {
    "major": 2,
    "minor": 4,
    "patch": 0
  },
  "pre_release": [
    "alpha-5-gcbc89373"
  ],
  "build": [
  ]
}
```

`v2.4.0-alpha.5-gcbc89373` means:

```json
{
  "version_core": { ...},
  "pre_release": [
    "alpha",
    "5-gcbc89373"
  ],
  "build": []
}
```

So exactly because we *don't* know what any pre-release ids on the tag mean, we should *not* combine those ids with our additional commit count/sha id.

It becomes even more obvious when the tag already contains multiple pre-release identifiers, like `v2.4.0-alpha.suse`:

`v2.4.0-alpha.suse-5-gcbc89373` means:

```json
{
  "version_core": { ...},
  "pre_release": [
    "alpha",
    "suse-5-gcbc89373"
  ],
  "build": []
}
```

Why does it make sense to combine the commit count/sha pre-release identifier with the `suse` id, but not with the `alpha` id?

So please reconsider and make it `v2.4.0-alpha.suse.5-gcbc89373` meaning

```json
{
  "version_core": { ...},
  "pre_release": [
    "alpha",
    "suse",
    "5-gcbc89373"
  ],
  "build": []
}
```

And just to address this pre-emptively: I would still consider the `-dirty` suffix to be a modifier of the commit id and not a separate pre-release id, so it should use a dash as a suffix separator:

`v2.4.0-alpha.suse.5-gcbc89373-dirty`, meaning:


```json
{
  "version_core": { ...},
  "pre_release": [
    "alpha",
    "suse",
    "5-gcbc89373-dirty"
  ],
  "build": []
}
```

I guess this means the regex I gave earlier to do the dash to dot conversion is incomplete and needs to account for the optional `-dirty` suffix. So we should have a test for this case. 😄

Reply:
I just realized that we may want to also separate the commit count and the commit sha into separate pre-release identifiers, so that they follow proper semver sorting order:

>Precedence for two pre-release versions with the same major, minor, and patch version MUST be determined by comparing each dot separated identifier from left to right until a difference is found as follows:
>
>Identifiers consisting of only digits are compared numerically.
>
>Identifiers with letters or hyphens are compared lexically in ASCII sort order.
>
>Numeric identifiers always have lower precedence than non-numeric identifiers.
>
>A larger set of pre-release fields has a higher precedence than a smaller set, if all of the preceding identifiers are equal.
>
>Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0.

The `<count>-g<sha1>` ids will always be compared lexically, so `v2.4.0-2-g1234567` will be considered later than `v2.4.0-19-gabcdef7` even though the latter has 17 additional commits. Switching to `v2.4.0-2.g1234567` etc. will still sort versions correctly based on commit count (but only if they live on the same branch).

I'm not sure if this is important to us, given that versions from e.g. different PRs inherently don't have a sort order until they are merged. It does properly sort main branch builds though, and it might come up even in CI builds from the same PR. If we do care, we should create versions like `v2.4.0-alpha.suse.5.gcbc89373-dirty`, meaning:

```json
{
  "version_core": { ...},
  "pre_release": [
    "alpha",
    "suse",
    "5",
    "gcbc89373-dirty"
  ],
  "build": []
}
```

**Off-topic** because only workflow related: sorting may be a good reason to add the "funny" tag (`v2.4.999`) on the main branch commit after the merge base for `release-2.4`, so all main line versions are semantically later than any builds from the `release-2.4` branch.

Reply:
Reading the semver spec again - you are correct @jandubois.
Dots are identifier separators, which is exactly what we want.
My thinking was that if the existing pre-release element contains dashes and dots already, that we'd run into trouble.
But semver is clear on this issue, and on sorting. So +1 on versions like these. For completeness, the example shows that the tag should be allowed to have dashes and dots.

```
1.2.3-my-alpha.build.15.gdeadbeef-dirty
```

Reply:
ping @thardeck